crates/blockchain-tree/src/chain.rs

```diff
@@ -178,7 +177,39 @@ impl AppendableChain {
 
     /// Validate and execute the given block that _extends the canonical chain_, validating its
     /// state root after execution.
-    fn validate_and_execute_canonical<PSDP, DB, C, EF>(
+    fn validate_and_execute<PSDP, DB, C, EF>(
+        block: SealedBlockWithSenders,
+        parent_block: &SealedHeader,
+        post_state_data_provider: PSDP,
+        externals: &TreeExternals<DB, C, EF>,
+        block_kind: BlockKind,
+    ) -> Result<PostState, Error>
+    where
+        PSDP: PostStateDataProvider,
+        DB: Database,
+        C: Consensus,
+        EF: ExecutorFactory,
+    {
+        if block_kind.extends_canonical_head() {
+            Self::validate_and_execute_canonical_head_descendant(
+                block,
+                parent_block,
+                post_state_data_provider,
+                externals,
+            )
+        } else {
+            Self::validate_and_execute_sidechain(
+                block,
+                parent_block,
+                post_state_data_provider,
+                externals,
+            )
```

Comment:
ah, yeah this is a nice way to require the caller to choose whether or not the block is descendant. instead of requiring the caller to choose the correct method